File: app/src/main/java/com/linderaredux/db/AppDbHelper.java
package com.linderaredux.db;

import com.linderaredux.api.response.patient.Patient;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;

public class AppDbHelper {

    private final AppDatabase mAppDatabase;

    public AppDbHelper(AppDatabase appDatabase) {
        this.mAppDatabase = appDatabase;
    }

    public Observable<List<Patient>> getAllPatients() {
        return Observable.fromCallable(() -> mAppDatabase.patientDao().loadAll());
    }

    public Observable<Boolean> savePatientList(List<Patient> patientList) {
        return Observable.fromCallable(() -> {
            mAppDatabase.patientDao().insertAll(patientList);
            return true;
        });
    }
}

File: app/src/main/java/com/linderaredux/db/AppDbHelper.kt
package com.linderaredux.db

import com.linderaredux.api.response.patient.Patient
import io.reactivex.Observable

class AppDbHelper(private val mAppDatabase: AppDatabase) {

    val allPatients: Observable<List<Patient>>
        get() = Observable.fromCallable { mAppDatabase.patientDao().loadAll() }

    fun savePatientList(patientList: List<Patient>): Observable<Boolean> {
        return Observable.fromCallable {
            patientList.forEach {
                insertOrUpdate(it)
            }
            true
        }
    }

    private fun insertOrUpdate(item: Patient) {
        mAppDatabase.runInTransaction {
            val id = mAppDatabase.patientDao().getItemId(item.id)
            if (id == null)
                mAppDatabase.patientDao().insert(item)
            else
                mAppDatabase.patientDao().update(item)
        }
    }
}